### PR TITLE
support extension-less page urls in deck

### PR DESCRIPTION
**Pull request by BenTheElder**
upstreaming this trick from my site so we have have https://prow.k8s.io/command-help instead of https://prow.k8s.io/command-help.html

Edit: refactored deck a bit and updated `runlocal` to use deck instead of python3, switched to extension-less URLs throughout deck.

Also fixed `cd`ing in runlocal so you can just `$ prow/cmd/deck/runlocal` without `cd`ing to deck first.

### CODE REVIEW

**Comment by BenTheElder** [comment]
I've been meaning to add this, WDYT all? Should be pretty esay to make the switch other than maybe the `runlocal` script.

**Comment by cjwagner** [comment]
👍 This is would be a nice change.
If it isn't easy to get this to work with `runlocal` it doesn't really matter. We can use paths with proper extensions while testing locally.

File: prow/cmd/deck/main.go [hunk 1]
```diff
@@ -147,6 +149,30 @@ func loadToken(file string) (string, error) {
...
+	r2.URL = new(url.URL)
+	*r2.URL = *original.URL
+	return r2
+}
+
+// serve with handler but map extensionless URLs to to the default
+func defaultExtension(extension string, h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if len(r.URL.Path) > 0 &&
+			r.URL.Path[len(r.URL.Path)-1] != '/' && path.Ext(r.URL.Path) == "" {
```
Comment (stevekuznetsov) [inline prow/cmd/deck/main.go]:
```go
path.Ext("somepath/") == ""
```

So the trailing slash check is superfluous

**Comment by stevekuznetsov** [comment]
This is cool, if the links within the app contain the extension, `runLocal` should be minimally impacted if you click around

File: prow/cmd/deck/main.go (hunk 1 above)
Comment (BenTheElder) [reply prow/cmd/deck/main.go]: This differentiates between `foo/` and `foo/bar`, so you can still serve directories with an index. we don't want to serve `foo/.html`...

Comment (BenTheElder) [reply prow/cmd/deck/main.go]: The idea is to only add a default extension to paths that look like serving a specific file like `/foo`, `/bar/baz` and not `/bar/` or `/`.

Comment (stevekuznetsov) [reply prow/cmd/deck/main.go]: Ah, right. ACK

**Comment by BenTheElder** [comment]
> This is cool, if the links within the app contain the extension, runLocal should be minimally impacted if you click around

I think I want to patch `runlocal` (maybe rewrite it in go?) so we can start using extensionless links in the app.

**Comment by cjwagner** [comment]
> I think I want to patch runlocal (maybe rewrite it in go?) so we can start using extensionless links in the app.

Should we just make it easier to run Deck locally? That would be helpful for testing front end changes that interact with the deck server.

**Comment by BenTheElder** [comment]
> Should we just make it easier to run Deck locally? That would be helpful for testing front end changes that interact with the deck server.

I for one find it useful to be able to develop these w/o depending on the network. Runlocal is nice because it's totally offline. We could do something inbetween though (reuse the file serving, but use a local copy of the api data)

**Comment by qhuynh96** [comment]
/lgtm
/hold
hold in case there is work in progress.

**Comment by kargakis** [comment]
Nice change!

**Comment by BenTheElder** [comment]
Added another commit that refactors `runlocal` and migrates the static content to extension-less. PTAL

File: prow/cmd/deck/main.go [hunk 2]
```diff
@@ -50,16 +51,42 @@ var (
 	redirectHTTPTo = flag.String("redirect-http-to", "", "host to redirect http->https to based on x-forwarded-proto == http.")
 	// use when behind an oauth proxy
 	hiddenOnly = flag.Bool("hidden-only", false, "Show only hidden jobs. Useful for serving hidden jobs behind an oauth proxy.")
+	runLocal   = flag.Bool("run-local", false,
+		"attempt to download data endpoints from --upstream and serve a local copy of the UI only")
```
Comment (kargakis) [inline prow/cmd/deck/main.go]: --upstream?

Comment (BenTheElder) [reply prow/cmd/deck/main.go]: Oops. At one point I was going to do the downloads in deck too but using curl works fine for now and having a wrapper makes it easy to run against the right $PWD. Will correct this, thanks.

Comment (BenTheElder) [reply prow/cmd/deck/main.go]: fixed

**Comment by BenTheElder** [comment]
Waiting until tomorrow morning (PST) for any more nits / thoughts then I'll `/hold cancel`. I think this is probably more or less fine now. The `prodOnlyMain` is kinda meh but.

File: prow/cmd/deck/main.go (hunk 2 above)
Comment (kargakis) [reply prow/cmd/deck/main.go]: +1 on doing the downloads as part of --run-local. Please open an issue to track the change.

File: prow/cmd/deck/main.go [hunk 3]
```diff
@@ -50,16 +51,41 @@ var (
 	redirectHTTPTo = flag.String("redirect-http-to", "", "host to redirect http->https to based on x-forwarded-proto == http.")
 	// use when behind an oauth proxy
 	hiddenOnly = flag.Bool("hidden-only", false, "Show only hidden jobs. Useful for serving hidden jobs behind an oauth proxy.")
+	runLocal   = flag.Bool("run-local", false, "serve a local copy of the UI, used by the prow/cmd/deck/runlocal script")
```
Comment (kargakis) [inline prow/cmd/deck/main.go]: nit: help messages usually start with a capital letter. Same comment about `redirect-http-to`, too.

Comment (BenTheElder) [reply prow/cmd/deck/main.go]: done!

File: prow/cmd/deck/main.go (hunk 2 above)
Comment (BenTheElder) [reply prow/cmd/deck/main.go]: opened one here: https://github.com/kubernetes/test-infra/issues/6698

**Pull request merged**